arctic/serialization/numpy_arrays.py

```diff
@@ -0,0 +1,179 @@
+import logging
+import numpy as np
+import numpy.ma as ma
+import pandas as pd
+
+from bson import Binary, SON
+
+from .._compress import compress, decompress
+
+
+class NumpyArrayConverter(object):
+    """
+    Converts a Numpy ndarray to and from PyMongo SON representation:
+
+        {
+         type:        '<i8',          # Numpy dtype (str)
+         length:      20,             # Numpy array length
+         values:      '\x00\x00bb...' # Compressed bytes
+        }
+    """
+
+    def docify(self, arr, meta=None):
+        """
+        Convert a Numpy ndarray to SON.
+
+        Parameters
+        ----------
+        arr:  ndarray
+            The numpy array to encode
+        meta: dict
+            Optional additional key value pairs to include in the SON
+        """
+        arr, mask = self._convert_types(arr)
+        son = SON(meta,
+                  type=arr.dtype.str,
+                  length=len(arr),
+                  values=Binary(compress(arr.tostring())))
+        if mask is not None:
+            son['mask'] = Binary(compress(mask.tostring()))
+        return son
+
+    def objify(self, doc):
+        """
+        Decode a Pymongo SON object into an Numpy ndarray
+        """
+        arr_data = decompress(doc['values'])
+        arr = np.fromstring(arr_data, doc['type'])
+
+        if 'mask' in doc:
+            mask_data = decompress(doc['mask'])
+            mask = np.fromstring(mask_data, 'bool')
+            arr = ma.masked_array(arr, mask)
+
+        return arr
+
+    def _convert_types(self, a):
+        """
+        Converts object arrays of strings to numpy string arrays
+        """
+        # No conversion for scalar type
+        if a.dtype != 'object':
+            return a, None
+
+        # We can't infer the type of an empty array, so just
+        # assume strings
+        if len(a) == 0:
+            return a.astype('U1'), None
+
+        # Compute a mask of missing values. Replace NaNs and Nones with
+        # empty strings so that type inference has a chance.
+        mask = pd.isnull(a)
+        if mask.sum() > 0:
+            a = a.copy()
+            np.putmask(a, mask, '')
+        else:
+            mask = None
+
+        if pd.lib.infer_dtype(a) == 'mixed':
+            a = np.array([s.encode('ascii') for s in a])
+            a = a.astype('O')
+
+        type_ = pd.lib.infer_dtype(a)
+        if type_ in ['unicode', 'string']:
+            max_len = pd.lib.max_len_string_array(a)
+            return a.astype('U{:d}'.format(max_len)), mask
+        else:
+            raise ValueError('Cannot store arrays with {} dtype'.format(type_))
+
+
+class FrameConverter(object):
+    """
+    Converts a Pandas Dataframe to and from PyMongo SON representation:
+
+        {
+         columns: [col1, col2, col3],
+         data: {
```

Comment:
Again this is so similar to:
https://github.com/manahl/arctic/blob/master/arctic/tickstore/tickstore.py#L36

Would be good to elide this